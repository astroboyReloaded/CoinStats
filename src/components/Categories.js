import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { fetchCategories } from '../redux/categories/categoriesSlice';

const Categories = () => {
  const { categories, isLoading, error } = useSelector(
    (state) => state.categories,
  );

  const dispatch = useDispatch();

  useEffect(() => {
    if (categories !== 'Categories') return;
    dispatch(fetchCategories());
  }, [dispatch, categories]);

  if (isLoading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Ups! Something went wrong</p>;
  }

  if (categories !== 'Categories') {
    return (
      <main>
        {categories.map((category) => (
          <article key={category.id}>
            <h1>{category.name}</h1>
            {category.top_3_coins.map((image) => (
              <img key={image} src={image} alt={image} />
            ))}
            <p>
              <data>{`Market Cap: $${category.market_cap}`}</data>
            </p>
            <p>
              <data>{`24h vlume: $ ${category.volume_24h}`}</data>
            </p>
            <p>
              <data>{`MC change 24h: $ ${category.market_cap_change_24h}`}</data>
            </p>
          </article>
        ))}
      </main>
    );
  }

  return categories;
};

export default Categories;
