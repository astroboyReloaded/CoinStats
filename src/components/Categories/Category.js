import { PropTypes } from 'prop-types';
import React from 'react';

const Category = ({
  name,
  top3,
  marketCap,
  vol24h,
  mrktCapChange24h,
}) => (
  <>
    <article className="category">
      <h1 className="category-name">{name}</h1>
      <div>
        {top3.map((image) => (
          <img key={Math.random()} src={image} alt={image} />
        ))}
      </div>
      <p className="categories-data-title">
        Market Cap:
      </p>
      <data className="categories-data">{`$${marketCap?.toLocaleString()}`}</data>
      <p className="categories-data-title">
        24h vlume:
      </p>
      <data className="categories-data">{`$${vol24h?.toLocaleString()}`}</data>
      <p className="categories-data-title">
        MC change 24h:
      </p>
      <data className={`categories-data ${mrktCapChange24h > 0 ? 'green' : 'red'}`}>{`${mrktCapChange24h?.toLocaleString().replace('-', '') || 'N/A'}%`}</data>
    </article>
  </>
);

Category.propTypes = {
  name: PropTypes.string.isRequired,
  top3: PropTypes.string.isRequired,
  marketCap: PropTypes.string.isRequired,
  vol24h: PropTypes.string.isRequired,
  mrktCapChange24h: PropTypes.string.isRequired,
};

export default Category;
