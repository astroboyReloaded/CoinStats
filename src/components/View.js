import { Outlet } from 'react-router-dom';
import Nav from './Nav';

const View = () => (
  <div className="wrapper">
    <Nav />
    <Outlet />
  </div>
);

export default View;
