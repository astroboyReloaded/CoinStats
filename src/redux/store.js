import { configureStore } from '@reduxjs/toolkit';
import allCoinsReducer from './all-coins/allCoinsSlice';
import coinDetailsReducer from './coin-details/coinDetailsSlice';
import categoriesReducer from './categories/categoriesSlice';
import globalDataReducer from './global-data/globalDataSlice';

const store = configureStore({
  reducer: {
    allCoins: allCoinsReducer,
    coinDetails: coinDetailsReducer,
    categories: categoriesReducer,
    globalData: globalDataReducer,
  },
});

export default store;
