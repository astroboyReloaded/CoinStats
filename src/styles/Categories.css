.categories-main {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.category {
  width: calc(50% - 0.6rem - 2px);
  margin: -1px -1px 0 0;
  border: 1px solid black;
  padding: 0.3rem;
  text-align: center;
}

.category-name {
  font-size: 1rem;
}

img {
  width: 2.1rem;
}

p {
  margin: 0;
}
