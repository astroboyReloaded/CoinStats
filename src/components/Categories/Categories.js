import useFetchCategories from '../../hooks/categories-hooks/useFetchCategories';
import '../../styles/Categories.css';
import Category from './Category';

const Categories = () => {
  const [categories, isLoading, error] = useFetchCategories();

  if (isLoading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Ups! Something went wrong</p>;
  }

  if (categories !== 'Categories') {
    return (
      <main className="categories-main">
        {categories.map((category) => (
          <Category
            key={category.id}
            name={category.name}
            top3={category.top_3_coins}
            marketCap={category.market_cap}
            vol24h={category.volume_24h}
            mrktCapChange24h={category.market_cap_change_24h}
          />
        ))}
      </main>
    );
  }

  return categories;
};

export default Categories;
