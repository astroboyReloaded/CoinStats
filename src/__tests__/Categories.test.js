import renderer from 'react-test-renderer';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';
import Categories from '../components/Categories/Categories.js';
import View from '../components/View';

it('renders Categories', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Provider store={store}>
          <Routes>
            <Route path="/" element={<View />}>
              <Route path="/categories" element={<Categories />} />
            </Route>
          </Routes>
        </Provider>
      </BrowserRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
