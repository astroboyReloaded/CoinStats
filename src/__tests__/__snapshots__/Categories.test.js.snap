// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Categories 1`] = `
<div
  className="wrapper"
>
  <nav>
    <img
      alt="gecko coin"
      className="logo"
      src="../icons/GeckoCoin-logo.jpg"
    />
    <h1
      className="app-title"
    >
      CoinStats
    </h1>
    <ul
      className="nav-ul"
    >
      <li>
        <a
          aria-current="page"
          className="navLink active"
          href="/"
          onClick={[Function]}
        >
          All Coins
        </a>
      </li>
      <li>
        <a
          className="navLink"
          href="/categories"
          onClick={[Function]}
        >
          Categories
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
